shared/js/background/https.es6.js

```diff
@@ -95,6 +95,67 @@ class HTTPS {
         // If it falls to here, default to reqUrl
         return reqUrl
     }
+
+    /**
+     * Modify response headers on https pages to fix mixed content.
+     * 1. For main frame requests, add uprade-insecure-requests directive
+     * 2. For subrequests, edit Access-Control-Allow-Origin header if it exists
+     *    to allow resources to be loaded on the https version of site.
+     */
+    setUpgradeInsecureRequestHeader (request) {
+        // skip requests to background tabs
+        if (request.tabId === -1) return {}
+
+        // Don't alter headers if site is whitelisted or http
+        if (request.url && request.url.indexOf('https://') === 0) {
+            let headersChanged = false
+            if (request.type === 'main_frame') {
+                let cspHeaderExists = false
+
+                for (const header in request.responseHeaders) {
+                    // If CSP header exists and doesn't include upgrade-insecure-request
+                    // directive, append it.
+                    if (request.responseHeaders[header].name.match(/Content-Security-Policy/i)) {
+                        cspHeaderExists = true
+                        const cspValue = request.responseHeaders[header].value
+
+                        if (!cspValue.match(/upgrade-insecure-requests/i)) {
+                            request.responseHeaders[header].value = 'upgrade-insecure-requests; ' + cspValue
+                            headersChanged = true
+                        }
+                    }
+                }
+                // If no CSP header, add one
+                if (!cspHeaderExists) {
+                    const upgradeInsecureRequests = {
+                        name: 'Content-Security-Policy',
+                        value: 'upgrade-insecure-requests'
+                    }
+                    request.responseHeaders.push(upgradeInsecureRequests)
+                    headersChanged = true
+                }
+            } else {
```

Comment:
Only need to edit access-control-allow-origin header on subrequests, since it doesn't really matter if the main document has this header.

Reply:
I can't think of any instances where this would cause breakage.. access-control-allow-origin should either be *, null or an url, and this will only kick in when it's an url and the main frame is https. In this case, this header needs to contain an https url in order for the resource to load.

It's possible that there are sites that we force upgrade to https that include the upgrade-insecure-requests header but haven't updated their resource request headers to reflect the new access-control-allow-origin value. This could be the case if a site has an https version of the site available but they primarily use the http version. Probably quite rare.

Reply:
Cool, let's leave it like that :+1: